opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp

```diff
@@ -613,8 +619,205 @@ class SimulatorFullyImplicitBlackoilEbos
             }
         }
     }
-
-
+    
+    //Fetch variables required for the well report
+    std::vector< std::vector<double>> getWellReport(const UnitSystem& units,
+                                   const Opm::PhaseUsage& pu,
+                                   const Wells* wells,
+                                   WellState& xw)
+    {
+        const int np = wells->number_of_phases;
+        int nw = wells->number_of_wells;
+        std::vector< std::vector<double>> wr ( 8 , std::vector<double> ( nw, 0.0 ) );
+        enum { Water = BlackoilPhases::Aqua, Oil = BlackoilPhases::Liquid, Gas = BlackoilPhases::Vapour };
+        
+        for (int w = 0; w < nw; ++w) {
+            double wellRate = 0.0;
+            const bool is_producer = wells->type[w] == PRODUCER;       
+            if( pu.phase_used[Water]) {
+                //Water phase rate per well
+                wr[0][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Water]]);
+            }              
+            if( pu.phase_used[Oil]) {
+                //Oil phase rate per well
+                wr[1][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Oil]]);
+            }         
+            if( pu.phase_used[Gas]) {
+                //Gas phase rate per well
+                wr[2][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Gas]]);
+            }
+            //Unit conversion
+            if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_FIELD) { 
+                wr[0][w] = unit::convert::to(wr[0][w],unit::stb/unit::day);
```

Comment:
I would assume we functionality to simplify this in opm-parser/opm-output yes. I will look and get back to you.